components/Ripple/src/MDCStatefulRippleView.m

```diff
@@ -0,0 +1,229 @@
+  self = [super initWithCoder:coder];
+  if (self) {
+    [self commonMDCStatefulRippleViewInit];
+  }
+  return self;
+}
+
+- (void)commonMDCStatefulRippleViewInit {
+  if (_rippleColors == nil) {
+    _rippleColors = [NSMutableDictionary dictionary];
+    UIColor *selectionColor = RippleSelectedColor();
+    _rippleColors[@(MDCRippleStateNormal)] = [UIColor colorWithWhite:0 alpha:kRippleAlpha];
+    _rippleColors[@(MDCRippleStateHighlighted)] = [UIColor colorWithWhite:0 alpha:kRippleAlpha];
+    _rippleColors[@(MDCRippleStateSelected)] =
+        [selectionColor colorWithAlphaComponent:kRippleSelectedAlpha];
+    _rippleColors[@(MDCRippleStateSelected | MDCRippleStateHighlighted)] =
+        [selectionColor colorWithAlphaComponent:kRippleAlpha];
+    _rippleColors[@(MDCRippleStateDragged)] = [UIColor colorWithWhite:0 alpha:kRippleDraggedAlpha];
+    _rippleColors[@(MDCRippleStateDragged | MDCRippleStateHighlighted)] =
+        [UIColor colorWithWhite:0 alpha:kRippleDraggedAlpha];
+    _rippleColors[@(MDCRippleStateSelected | MDCRippleStateDragged)] =
+        [selectionColor colorWithAlphaComponent:kRippleDraggedAlpha];
+  }
+  self.userInteractionEnabled = YES;
+}
+
+- (UIColor *)rippleColorForState:(MDCRippleState)state {
+  UIColor *rippleColor = _rippleColors[@(state)];
+  if (rippleColor == nil && (state & MDCRippleStateDragged) != 0) {
+    rippleColor = _rippleColors[@(MDCRippleStateDragged)];
+  } else if (rippleColor == nil && (state & MDCRippleStateSelected) != 0) {
+    rippleColor = _rippleColors[@(MDCRippleStateSelected)];
+  }
+
+  if (rippleColor == nil) {
+    rippleColor = _rippleColors[@(MDCRippleStateNormal)];
+  }
+  return rippleColor;
+}
+
+- (void)updateRippleColor {
+  UIColor *rippleColor = [self rippleColorForState:self.state];
+  [self setRippleColor:rippleColor];
+}
+
+- (void)updateActiveRippleColor {
+  UIColor *rippleColor = [self rippleColorForState:self.state];
+  [self setActiveRippleColor:rippleColor];
+}
+
+- (void)setRippleColor:(UIColor *)rippleColor forState:(MDCRippleState)state {
+  _rippleColors[@(state)] = rippleColor;
+
+  [self updateRippleColor];
+}
+
+- (MDCRippleState)state {
+  NSInteger state = 0;
+  if (self.selected) {
+    state |= MDCRippleStateSelected;
+  }
+  if (self.rippleHighlighted) {
+    state |= MDCRippleStateHighlighted;
+  }
+  if (self.dragged) {
+    state |= MDCRippleStateDragged;
+  }
+  return state;
+}
+
+- (void)setSelected:(BOOL)selected {
+  if (!self.allowsSelection || _tapWentOutsideOfBounds) {
+    // If we disallow selection we don't want to apply any visual or state changes for selection.
+    return;
+  }
+  if (selected == _selected && self.activeRippleLayer) {
+    return;
+  }
+  _selected = selected;
+  // Go into the selected state visually.
+  if (selected) {
+    if (!self.activeRippleLayer) {
+      // If we go into the selected state manually, without coming from the highlighted state,
+      // We present the ripple overlay instantly without animation.
```

Comment:
This comment implies that the highlighted property should come in to play here: is activeRippleLayer a proxy for that state?

It may improve readability here to either check the highlighted state as the comment implies or to explain why activeRippleLayer is a proxy for that state.

Reply:
The way cells work in UIKit is that the highlighted state gets set to NO before the selected state is set to YES, therefore if I check the highlighted state here, it wouldn't work in the cell's case. A more precise check is to see if there is already an active overlay and decide if there is a need for another one or not. I updated the inline comment to be more verbose.